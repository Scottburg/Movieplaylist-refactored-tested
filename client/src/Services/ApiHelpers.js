export default {
  fetchRequest: (url, options) => {
    return fetch(url, options)
      .then((res) => (res.status <= 400 ? res : Promise.reject(res)))
      .then((res) => res.json())
      .catch((err) => {
        console.log(err);
      });
  },
};
