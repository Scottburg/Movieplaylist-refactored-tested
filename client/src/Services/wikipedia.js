const wtf = require('wtf_wikipedia');
const baseUrl = 'https://en.wikipedia.org/w/api.php';

export function getWikipedia(title) {
  const params = {
    action: 'opensearch',
    search: title,
    limit: '50',
    namespace: '0',
    format: 'json',
  };

  let url = baseUrl + '?origin=*';
  Object.keys(params).forEach(function (key) {
    url += '&' + key + '=' + params[key];
  });

  const conditions = ['soundtrack', 'music', 'OST', 'Music'];
  return fetch(url)
    .then((res) => res.json())
    .then((res) => {
      const wikiUrls = res[3];
      for (let wikiUrl of wikiUrls) {
        if (conditions.some((el) => wikiUrl.includes(el))) {
          return wikiUrl;
        }
      }
    })
    .catch(/*Put catch code*/);
}

export async function getSongList(title) {
  const result = {
    titles: [],
    artists: [],
  };
  const wikiUrl = await getWikipedia(title);
  if (wikiUrl) {
    return wtf
      .fetch(wikiUrl)
      .then((doc) => {
        return doc.json();
      })
      .then((doc) => {
        for (let section of doc.sections) {
          if (section.title === 'Track listing') return section;
        }
      })
      .then((data) => {
        if (data?.templates) {
          return data.templates[0];
        } else return undefined;
      })
      .then((data) => {
        for (let key in data) {
          if (key.includes('title')) result.titles.push(data[key]);
          if (key.includes('extra')) result.artists.push(data[key]);
        }

        return result;
      });
  }
  return result;
}
