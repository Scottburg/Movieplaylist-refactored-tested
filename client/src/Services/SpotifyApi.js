import ApiHelpers from './ApiHelpers';

const BASE_URL = 'https://api.spotify.com/v1';

export default {
  getSpotifyUserId: async (token) => {
    const options = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    return await ApiHelpers.fetchRequest(`${BASE_URL}/me`, options);
  },

  createPlaylist: async (userId, movieTitle, token) => {
    const url = `${BASE_URL}/users/${userId}/playlists`;
    const bodyOption = {
      name: `${movieTitle} SoundTrack`,
      public: false,
      description: 'Create by Movie Playlist Generator',
    };
    const options = {
      method: 'POST',
      mode: 'cors',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      Accept: 'application/json',
      body: JSON.stringify(bodyOption),
    };
    return await ApiHelpers.fetchRequest(url, options);
  },

  searchAlbum: async (movieTitle, token) => {
    const url = `${BASE_URL}/search` +
      `?q=album:${movieTitle}+soundtrack` +
      `&type=album,playlist` +
      `&market=from_token`;
    const options = {
      method: 'GET',
      mode: 'cors',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      Accept: 'application/json',
    };
    let songs = [];
    try {
      const res = await ApiHelpers.fetchRequest(url, options);
      if (res.albums && res.albums.items && res.albums.items[0]) {
        const album = res.albums.items[0];
        const albumUrl = `${BASE_URL}/albums/${album.id}/tracks` +
        `?limit=50`;
        const songsRes = await ApiHelpers.fetchRequest(albumUrl, options);
        if (songsRes) songs = songsRes.items;
      } else if (res.playlists && res.playlists.items && res.playlists.items[0]) {
        const playlist = res.playlists.items[0];
        const playlistUrl = `${BASE_URL}/playlists/${playlist.id}/tracks` +
        `?limit=50`;
        const songsRes = await ApiHelpers.fetchRequest(playlistUrl, options);
        if (songsRes) songs = songsRes.items;
      }
    } catch (err) {
      return [];
    }
    return songs.map(song => ({
      song: song.name,
      artist: song.artists.map(artist => artist.name).join(', '),
      id: song.id
    }));
  },

  searchSongs: async (songs, token) => {
    if (!songs) return;
    async function asyncForEach(array, callback) {
      for (let index = 0; index < array.length; index++) {
        await callback(array[index], index, array);
      }
    }
    const result = [];
    await asyncForEach(songs, async (song) => {
      let url = '';
      if (song.artist) {
        url = `${BASE_URL}/search?q=${encodeURIComponent(
          song.song
        )}%20artist:${encodeURIComponent(song.artist)}&type=track&limit=10`;
      } else {
        url = `${BASE_URL}/search?q=${encodeURIComponent(
          song.song
        )}&type=track&limit=10`;
      }
      const response = await fetch(url, {
        method: 'GET',
        mode: 'cors',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        Accept: 'application/json',
      });
      const res = await response.json();
      const id = res.tracks.items[0] !== undefined && res.tracks.items[0].id;
      id && result.push(id);
    });
    return result;
  },

  addSongs: async (songsId, playlistId, token) => {
    if (!songsId) return;
    let queryWithoutId = 'spotify:track:';
    let query = songsId.map((id) => `${queryWithoutId + id}`).join();
    const url = `${BASE_URL}/playlists/${playlistId}/tracks?uris=${encodeURIComponent(
      query
    )}`;
    const options = {
      method: 'POST',
      mode: 'cors',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      Accept: 'application/json',
    };
    return await ApiHelpers.fetchRequest(url, options);
  },
};
