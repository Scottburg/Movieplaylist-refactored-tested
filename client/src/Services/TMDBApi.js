import ApiHelpers from './ApiHelpers';

const BASE_URL = 'https://api.themoviedb.org/3';

export function getMovieList(keyword) {
  return ApiHelpers.fetchRequest(
    `${BASE_URL}/search/movie?api_key=${
      process.env.REACT_APP_TMDB_ID
    }&language=en-US&query=${encodeURIComponent(
      keyword
    )}&page=1&include_adult=false`
  );
}