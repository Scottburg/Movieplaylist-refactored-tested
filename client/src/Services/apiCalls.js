import ApiHelpers from './ApiHelpers';

export function getMovieList(keyword) {
  return ApiHelpers.fetchRequest(
    `https://api.themoviedb.org/3/search/movie?api_key=${
      process.env.REACT_APP_TMDB_ID
    }&language=en-US&query=${encodeURIComponent(
      keyword
    )}&page=1&include_adult=false`
  );
}

export function getSpotifyUserId(token) {
  const options = {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  };
  return ApiHelpers.fetchRequest('https://api.spotify.com/v1/me', options);
}
