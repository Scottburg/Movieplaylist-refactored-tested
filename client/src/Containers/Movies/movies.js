import React, { useState, useEffect } from 'react';
import MovieItem from '../../Components/MovieItem/MovieItem';
import { getMovieList } from '../../Services/apiCalls';
import './movies.css';

function Movies({ searching, triggerSearch }) {
  const [movies, setMovies] = useState();

  useEffect(() => {
    getMovieList(searching).then((data) => setMovies(data));
  }, [triggerSearch]);

  return (
    <div>
      <div className="movies">
        {movies &&
          movies.results
            .slice(0, 10)
            .map((movie) => (
              <MovieItem
                key={movie.id}
                title={movie.title}
                posterPath={movie.poster_path}
                date={movie.release_date}
              />
            ))}
      </div>
      <p className="no-movies">
        No more movies, if your movie is not in the list try another search
      </p>
    </div>
  );
}

export default Movies;
