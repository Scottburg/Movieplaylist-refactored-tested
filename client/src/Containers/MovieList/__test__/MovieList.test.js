import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react';

import Movies from '../MovieList';

// mocking of apiCalls
import { getMovieList } from '../../../Services/apiCalls';
jest.mock('../../../Services/apiCalls');
const fakeMoviesPromise = Promise.resolve({
  results: [
    {
      id: '27205',
      title: 'Inception',
      poster_path: '/edv5CZvWj09upOsy2Y6IwDhK8bt.jpg',
      release_date: '2010-12-07',
    },
    {
      id: '597',
      title: 'Titanic',
      poster_path: '/9xjZS2rlVxm8SFx8kPC3aIGCOYQ.jpg',
      release_date: '1997-11-18',
    },
  ],
});
getMovieList.mockImplementation(() => fakeMoviesPromise);

describe('Movie component', () => {
  afterEach(cleanup);
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Movies />, div);
  });
  it('renders a footer message', () => {
    const component = render(<Movies />);
    expect(component.getByText(/No more movies/i)).toBeInTheDocument();
  });
  it('renders each movie occurence as a SingleMovie component', async () => {
    const { container } = await render(<Movies />);
    expect(container.getElementsByClassName('singleMovie').length).toEqual(
      (await fakeMoviesPromise).results.length
    );
  });
  it('renders the movie titles', async () => {
    const { queryByText } = await render(<Movies />);
    expect(queryByText(/Inception/i)).toBeInTheDocument();
    expect(queryByText(/Titanic/i)).toBeInTheDocument();
  });
  it('should match snapshot', () => {
    const component = renderer.create(<Movies />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
