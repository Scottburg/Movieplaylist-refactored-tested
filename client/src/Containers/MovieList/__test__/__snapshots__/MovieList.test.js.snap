// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie component should match snapshot (when there are movies) 1`] = `
<div>
  <div
    className="movies"
  >
    <div
      className="singleMovie"
    >
      <div>
        <span>
          <h5
            data-testid="title"
          >
            Inception
          </h5>
        </span>
        <img
          alt="movie post"
          className="poster-img"
          src="http://image.tmdb.org/t/p/w200/edv5CZvWj09upOsy2Y6IwDhK8bt.jpg"
          style={
            Object {
              "height": "300px",
              "width": "200px",
            }
          }
        />
        <button
          className="playlistButton btn btn-outline-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Show Playlist
        </button>
      </div>
    </div>
    <div
      className="singleMovie"
    >
      <div>
        <span>
          <h5
            data-testid="title"
          >
            Titanic
          </h5>
        </span>
        <img
          alt="movie post"
          className="poster-img"
          src="http://image.tmdb.org/t/p/w200/9xjZS2rlVxm8SFx8kPC3aIGCOYQ.jpg"
          style={
            Object {
              "height": "300px",
              "width": "200px",
            }
          }
        />
        <button
          className="playlistButton btn btn-outline-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Show Playlist
        </button>
      </div>
    </div>
  </div>
  <p
    className="no-movies"
  >
    No more movies, if your movie is not in the list try another search
  </p>
</div>
`;

exports[`Movie component should match snapshot (when there are no movies) 1`] = `
<div>
  <div
    className="movies"
  />
  <p
    className="no-movies"
  >
    No more movies, if your movie is not in the list try another search
  </p>
</div>
`;
